Tests/SwiftFusionTests/Inference/NonlinearFactorGraphTests.swift

```diff
@@ -0,0 +1,99 @@
+import SwiftFusion
+import TensorFlow
+import XCTest
+
+final class NonlinearFactorGraphTests: XCTestCase {
+  /// test ATr
+  func testBasicOps() {
+    var fg = NonlinearFactorGraph()
+    
+    let bf1 = BetweenFactor(0, 1, Pose2(0.0,0.0, 0.0))
+    
+    fg += AnyNonlinearFactor(bf1)
+    
+    var val = Values()
+    val.insert(0, AnyDifferentiable(Pose2(1.0, 1.0, 0.0)))
+    val.insert(1, AnyDifferentiable(Pose2(1.0, 1.0, .pi)))
+    
+    let gfg = fg.linearize(val)
+    
+    var vv = VectorValues()
+    
+    vv.insert(0, Tensor<Double>(shape:[3, 1], scalars: [0.0, 0.0, 0.0]))
+    vv.insert(1, Tensor<Double>(shape:[3, 1], scalars: [0.0, 0.0, 0.0]))
+    
+    let expected = Tensor<Double>(shape:[3, 1], scalars: [.pi, 0.0, 0.0])
+    
+    print("gfg = \(gfg)")
+    print("error = \(gfg.residual(vv).norm)")
+    assertEqual((gfg.residual(vv))[0], expected, accuracy: 1e-6)
+  }
+  
+  /// test CGLS iterative solver
+  func testCGLSPose2SLAM() {
+    // Initial estimate for poses
+    let p1T0 = Pose2(Rot2(0.2), Vector2(0.5, 0.0))
+    let p2T0 = Pose2(Rot2(-0.2), Vector2(2.3, 0.1))
+    let p3T0 = Pose2(Rot2(.pi / 2), Vector2(4.1, 0.1))
+    let p4T0 = Pose2(Rot2(.pi), Vector2(4.0, 2.0))
+    let p5T0 = Pose2(Rot2(-.pi / 2), Vector2(2.1, 2.1))
+
+    let map = [p1T0, p2T0, p3T0, p4T0, p5T0]
+
+    var fg = NonlinearFactorGraph()
+    
+    fg += AnyNonlinearFactor(BetweenFactor(1, 0, Pose2(2.0, 0.0, .pi / 2)))
+    fg += AnyNonlinearFactor(BetweenFactor(2, 1, Pose2(2.0, 0.0, .pi / 2)))
+    fg += AnyNonlinearFactor(BetweenFactor(3, 2, Pose2(2.0, 0.0, .pi / 2)))
+    fg += AnyNonlinearFactor(BetweenFactor(4, 3, Pose2(2.0, 0.0, .pi / 2)))
+    fg += AnyNonlinearFactor(PriorFactor(0, Pose2(0.0, 0.0, 0.0)))
```

Comment:
I'd love to be able to just say `fg += BetweenFactor(...)`

Reply:
Deleting `AnyNonlinearFactor` would make it like that!

Reply:
Done :)